src/backend/gporca/libgpopt/src/operators/CLogicalReturning.cpp

```diff
@@ -0,0 +1,203 @@
+CLogicalReturning::CLogicalReturning(CMemoryPool *mp,
+									 CTableDescriptor *ptabdesc)
+	: CLogical(mp), m_ptabdesc(ptabdesc)
+{
+	GPOS_ASSERT(NULL != ptabdesc);
+
+	m_pdrgpcrOutput =
+		PdrgpcrCreateMapping(mp, ptabdesc->Pdrgpcoldesc(), UlOpId());
+
+	m_pcrsLocalUsed->Include(m_pdrgpcrOutput);
+}
+
+//---------------------------------------------------------------------------
+//	@function:
+//		CCLogicalReturning::CLogicalReturning
+//
+//	@doc:
+//		Ctor
+//
+//---------------------------------------------------------------------------
+CLogicalReturning::CLogicalReturning(CMemoryPool *mp,
+									 CTableDescriptor *ptabdesc,
+									 CColRefArray *pdrgpcrOutput)
+	: CLogical(mp), m_ptabdesc(ptabdesc), m_pdrgpcrOutput(pdrgpcrOutput)
+{
+	GPOS_ASSERT(NULL != ptabdesc);
+	GPOS_ASSERT(NULL != pdrgpcrOutput);
+
+	m_pcrsLocalUsed->Include(m_pdrgpcrOutput);
+}
+
+//---------------------------------------------------------------------------
+//	@function:
+//		CLogicalReturning::~CLogicalReturning
+//
+//	@doc:
+//		Dtor
+//
+//---------------------------------------------------------------------------
+CLogicalReturning::~CLogicalReturning()
+{
+	CRefCount::SafeRelease(m_ptabdesc);
+	CRefCount::SafeRelease(m_pdrgpcrOutput);
+}
+
+//---------------------------------------------------------------------------
+//	@function:
+//		CLogicalReturning::Matches
+//
+//	@doc:
+//		Match function
+//
+//---------------------------------------------------------------------------
+BOOL
+CLogicalReturning::MatchesReturning(CLogicalReturning *popReturning) const
+{
+	return m_ptabdesc->MDId()->Equals(popReturning->Ptabdesc()->MDId()) &&
+		   m_pdrgpcrOutput->Equals(popReturning->PdrgpcrOutput());
+}
+
+//---------------------------------------------------------------------------
+//	@function:
+//		CLogicalReturning::HashValue
+//
+//	@doc:
+//		Hash function
+//
+//---------------------------------------------------------------------------
+ULONG
+CLogicalReturning::HashValue() const
+{
+	ULONG ulHash = gpos::CombineHashes(COperator::HashValue(),
+									   m_ptabdesc->MDId()->HashValue());
+
+	ulHash =
+		gpos::CombineHashes(ulHash, CUtils::UlHashColArray(m_pdrgpcrOutput));
+
+	return ulHash;
+}
+
+//---------------------------------------------------------------------------
+//	@function:
+//		CLogicalReturning::CopyRemappedColumns
+//
+//	@doc:
+//		return a copy of output columns
+//
+//---------------------------------------------------------------------------
+CColRefArray *
+CLogicalReturning::CopyRemappedColumns(CMemoryPool *mp,
+									   UlongToColRefMap *colref_mapping,
+									   BOOL must_exist)
+{
+	CColRefArray *pdrgpcrOutput = NULL;
+	if (must_exist)
+	{
+		pdrgpcrOutput =
+			CUtils::PdrgpcrRemapAndCreate(mp, m_pdrgpcrOutput, colref_mapping);
+	}
+	else
+	{
+		pdrgpcrOutput = CUtils::PdrgpcrRemap(mp, m_pdrgpcrOutput,
+											 colref_mapping, must_exist);
+	}
+
+	return pdrgpcrOutput;
```

Comment:
What about don't assing NULL and using ternary operator to emphasise that the return value of the both functions is assinged to one variable? I think that the pdrgpcrOutput is redundant and the code will be simpler without it
```suggestion
	return (must_exist) ? CUtils::PdrgpcrRemapAndCreate(mp, m_pdrgpcrOutput,
														colref_mapping)
						: CUtils::PdrgpcrRemap(mp, m_pdrgpcrOutput,
											   colref_mapping, must_exist);
```